File: Data Structures & Algorithms/two-integer-sum/submission-2.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> diff;
        for (int i = 0; i < nums.size(); ++i) {
            int difference = target - nums[i];
            diff[difference] = i;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (diff.find(nums[i]) != diff.end()) {
                if (i == diff[nums[i]]) {
                    continue;
                } else if (i < diff[nums[i]]) {
                    return {i, diff[nums[i]]};
                } else {
                    return {diff[nums[i]], i};
                }
            }
        }
    }
};

File: Data Structures & Algorithms/find-duplicate-integer/submission-2.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        s, f = 0, 0
        
        while True:
            s = nums[s] # move 1
            f = nums[nums[f]] # move 2
            if s == f:
                break
        
        i, j = 0, s
        while i != j:
            i = nums[i]
            j = nums[j]
        
        return i